reviewer/reviewer.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reviewer
+
+import (
+	"context"
+	"math/rand"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+
+	"github.com/palantir/policy-bot/policy/common"
+	"github.com/palantir/policy-bot/pull"
+)
+
+func findLeafChildren(result common.Result) []common.Result {
+	var r []common.Result
+	if len(result.Children) == 0 {
+		if result.Status == common.StatusPending && result.Error == nil {
+			return []common.Result{result}
+		}
+	} else {
+		for _, c := range result.Children {
+			if c.Status == common.StatusPending {
+				r = append(r, findLeafChildren(*c)...)
+			}
+		}
+	}
+	return r
+}
+
+// select n random values from the list of users without reuse
+func selectRandomUsers(n int, users []string, r *rand.Rand) []string {
+	selected := make(map[int]bool)
+
+	var selections []string
+	if n == 0 {
+		return selections
+	}
+	if n >= len(users) {
+		return users
+	}
+
+	for i := 0; i < n; i++ {
+		for {
+			i := r.Intn(len(users))
+			if !selected[i] {
+				selected[i] = true
+				selections = append(selections, users[i])
+				break
+			}
+		}
+	}
+	return selections
+}
+
+func shoveIntoMap(m map[string]struct{}, u []string) map[string]struct{} {
+	for _, n := range u {
+		m[n] = struct{}{}
+	}
+	return m
+}
+
+func selectTeamMembers(prctx pull.Context, allTeams []string, r *rand.Rand) ([]string, error) {
+	randomTeam := allTeams[r.Intn(len(allTeams))]
+	teamInfo := strings.Split(randomTeam, "/")
+	teamMembers, err := prctx.ListTeamMembers(teamInfo[0], teamInfo[1])
+	if err != nil {
+		return nil, errors.Wrapf(err, "Unable to get member listing for team %s", randomTeam)
+	}
+	return teamMembers, nil
+}
+
+func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common.Result, r *rand.Rand) ([]string, error) {
+	logger := zerolog.Ctx(ctx)
+	pendingLeafNodes := findLeafChildren(result)
+	var requestedUsers []string
+
+	logger.Debug().Msgf("Collecting reviewers for %d pending leaf nodes", len(pendingLeafNodes))
+
+	for _, child := range pendingLeafNodes {
+		allUsers := make(map[string]struct{})
+		allUsers = shoveIntoMap(allUsers, child.Rule.RequestedUsers)
+
+		if len(child.Rule.RequestedTeams) > 0 {
+			teamMembers, err := selectTeamMembers(prctx, child.Rule.RequestedTeams, r)
```

Comment:
Have we filtered out non-organization teams prior to this point? Or is the idea that we should consider all teams, because they might have members who are collaborators on this project?

If it's the later, that's even more reason to define the listing methods on the `MembershipContext`: so you have permissions to list teams and members of different organizations.

Reply:
We haven't filtered anything, for exactly the reason you suggested. We could either filter out non-org teams, but I think it makes more sense to move these into the membership context methods.